src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -5646,9 +5679,12 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
             // scala/bug#5967 Important to replace param type A* with Seq[A] when seen from from a reference, to avoid
             //         inference errors in pattern matching.
               stabilize(tree2, pre2, mode, pt) modifyType dropIllegalStarTypes
-            }) setAttachments tree.attachments
-          }
+            }
+            if (!isPastTyper && currentRun.isScala3 && !isScala3cross && isFunctionType(pt) && sym.isModule && sym.isSynthetic && sym.companion.isCase)
+              context.deprecationWarning(tree.pos, sym, s"Synthetic case companion used as a Function, use explicit object with Function parent", "2.13.13")
```

Comment:
The message could suggest `-Xsource:3cross`? The end goal also in Scala 3 is explicit, either explicit parent or explicit `C.apply`. 🤔